loadfile/loadfile.go

```diff
@@ -7,41 +7,168 @@ import (
 	}
-	requiredFilesAbs := map[string]string{}
-	for key, f := range requiredFiles {
+	filesAbsPaths := map[string]ContextFile{}
+	for key, f := range files {
 		abspath := f
 		if !filepath.IsAbs(f) {
 			abspath = filepath.Join(absDir, f)
 		}
-		requiredFilesAbs[key] = abspath
+		filesAbsPaths[key] = ContextFile{
+			ID:           f,
+			AbsolutePath: abspath,
+		}
+	}
+	return &fileCache{
+		rootDir: absDir, files: filesAbsPaths}, nil
+}
+
+// NewFileCache creates a file cache where the fileContents is
+// a mapping of paths to file contents. The keys in fileContents
+// become the keys in Files, and the absolute paths and ID for
+// each context file.
+func NewFileCache(rootDir string, fileContents map[string][]byte) FileCache {
+	files := map[string]ContextFile{}
+	for key, content := range fileContents {
+		files[key] = ContextFile{
+			ID:           key,
+			AbsolutePath: key,
+			Content:      content,
+		}
+	}
+	return &fileCache{
+		rootDir: rootDir,
+		files:   files,
+	}
+}
+
+// LoadContext reads the content of each context file into Content.
+func (fc *fileCache) LoadContext() error {
+	result := map[string]ContextFile{}
+	var err error
+	for key, cf := range fc.files {
+		absPath := cf.AbsolutePath
+		fileData, err := os.ReadFile(filepath.Clean(absPath))
+		if err != nil {
+			return fmt.Errorf("error reading file %s (%w)", absPath, err)
+		}
+		result[key] = ContextFile{
+			ID:           cf.ID,
+			AbsolutePath: cf.AbsolutePath,
+			Content:      fileData,
+		}
+	}
+	fc.files = result
+	return err
+}
+
+// RootDir returns the root directory used by files with
+// relative file paths.
+func (fc *fileCache) RootDir() string {
+	return fc.rootDir
+}
+
+// GetByKey asks if the file cache contains the given fileKey.
+func (fc *fileCache) GetByKey(fileKey string) (*ContextFile, bool) {
+	cf, ok := fc.files[fileKey]
+	return &cf, ok
+}
+
+// AbsPathByKey returns the absolute file path of a given file key,
+// if it exists in the file cache, nil otherwise.
+func (fc *fileCache) AbsPathByKey(fileKey string) *string {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return &cf.AbsolutePath
+}
+
+// ContentByKey returns the file content of a given file key, if it
+// exists in the file cache, nil otherwise.
+func (fc *fileCache) ContentByKey(fileKey string) []byte {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return cf.Content
+}
+
+// Contents returns a mapping of the file cache's file keys to file Content.
+func (fc *fileCache) Contents() map[string][]byte {
+	result := map[string][]byte{}
+	for key, f := range fc.files {
+		result[key] = f.Content
+	}
+	return result
+}
+
+// AbsPaths returns a mapping of the file cache's file keys to
+// absolute file paths.
+func (fc *fileCache) AbsPaths() map[string]string {
+	result := map[string]string{}
+	for key, f := range fc.files {
+		result[key] = f.AbsolutePath
+	}
+	return result
+}
```

Comment:
Why build a new map here (`Contents` and `AbsPaths`) instead of just looking up the key and fetching the field of the `ContextFile` like `AbsPathByKey`? Are these really needed? Making "minimized" copies of the original and potentially passing them around seems to violate "single truth" principles, and I don't see offhand why there would be an advantage. 😕

Reply:
At this point, it was mostly a convenience function for testing. Should it be taken out?

Reply:
`Contents()` is currently used (outside of the tests).  I don't see a similar justification for `AbsPaths()`.

Reply:
I'll rewrite it in `loadfile_test` .